pkg/processor/schemaregistry/avro/schema.go

```diff
@@ -0,0 +1,202 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package avro
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/hamba/avro/v2"
+)
+
+// ExtractSchema uses reflection to extract an Avro schema from v.
+func ExtractSchema(v any) (avro.Schema, error) {
+	return reflectInternal([]string{"record"}, reflect.ValueOf(v), reflect.TypeOf(v))
+}
+
+// SortFields is a utility for tests to ensure the schemas can be compared.
+func SortFields(s avro.Schema) {
+	rs, ok := s.(*avro.RecordSchema)
+	if !ok {
+		return
+	}
+	fields := rs.Fields()
+	sort.Slice(fields, func(i, j int) bool {
+		return fields[i].Name() < fields[j].Name()
+	})
+	for i := range fields {
+		SortFields(fields[i].Type())
+	}
+}
+
+//nolint:gocyclo // reflection requires a huge switch, it's fine
+func reflectInternal(path []string, v reflect.Value, t reflect.Type) (avro.Schema, error) {
+	if t == nil {
+		return nil, nil // untyped nil
+	}
+	switch t.Kind() {
+	case reflect.Bool:
+		return avro.NewPrimitiveSchema(avro.Boolean, nil), nil
+	case reflect.Int, reflect.Int64, reflect.Uint32:
+		return avro.NewPrimitiveSchema(avro.Long, nil), nil
+	case reflect.Int32, reflect.Int16, reflect.Uint16, reflect.Int8, reflect.Uint8:
+		return avro.NewPrimitiveSchema(avro.Int, nil), nil
+	case reflect.Float32:
+		return avro.NewPrimitiveSchema(avro.Float, nil), nil
+	case reflect.Float64:
+		return avro.NewPrimitiveSchema(avro.Double, nil), nil
+	case reflect.String:
+		return avro.NewPrimitiveSchema(avro.String, nil), nil
+	case reflect.Pointer:
+		s, err := reflectInternal(path, v.Elem(), t.Elem())
+		if err != nil {
+			return nil, err
+		}
+
+		var schemas avro.Schemas
+		if us, ok := s.(*avro.UnionSchema); ok && us.Nullable() {
+			// it's already a nullable schema
+			return s, nil
+		} else if ok {
+			// take types from union schema
+			schemas = us.Types()
+		} else if s.Type() != avro.Null {
+			// non-nil type
+			schemas = avro.Schemas{s}
+		}
+
+		s, err = avro.NewUnionSchema(append(schemas, &avro.NullSchema{}))
+		if err != nil {
+			return nil, err
+		}
+
+		return s, nil
+	case reflect.Interface:
+		if !v.IsValid() || v.IsNil() {
```

Comment:
`IsValid` will be `false` when `reflect.Value` is the zero value (e.g. `reflect.Value{}`). We sometimes pass down the zero value, for example when we are trying to extract the schema of a slice with a concrete type (e.g. `[]int`). in that case we ignore the actual values in the schema and simply rely on the type information of the slice. If that type is a struct (e.g. `[]MyStruct`) we will further traverse the fields and have no value to send alongside the field, so it can happen that the value is invalid.